### PR TITLE
remove exprt::negate, sum, mul, subtract (deprecated since 2011)

### PR DESCRIPTION
**Pull request by kroening**
Note that mul_expr and sum_expr could use modernisation -- however, I'd say that this is a separate PR.

### CODE REVIEW

File: src/util/simplify_expr_int.cpp [hunk 1]
```diff
@@ -53,6 +54,114 @@ bool simplify_exprt::simplify_bswap(bswap_exprt &expr)
...
+      dest.set(ID_value, a_plus_b.get_string(ID_value));
+      return false;
+    }
+  }
+  else if(type_id==ID_unsignedbv || type_id==ID_signedbv)
+  {
+    dest.set(ID_value, integer2binary(
+      binary2integer(dest.get_string(ID_value), false)+
+      binary2integer(expr.get_string(ID_value), false),
+      unsafe_string2unsigned(dest.type().get_string(ID_width))));
```
Comment (tautschnig) [inline src/util/simplify_expr_int.cpp]: Use `to_bitvector_type(dest.type()).get_width()`

File: src/util/guard.h [hunk 2]
```diff
@@ -14,14 +14,14 @@ Author: Daniel Kroening, kroening@kroening.com
...
 
-#include "expr.h"
+#include "std_expr.h"
 
 class guardt:public exprt
 {
 public:
   guardt()
   {
-    make_true();
```
Comment (tautschnig) [inline src/util/guard.h]: This seems unrelated? (I accept that it might be a similar piece of cleanup, but this is not covered by the commit description.)

File: src/util/simplify_expr_int.cpp [hunk 3]
```diff
@@ -53,6 +54,114 @@ bool simplify_exprt::simplify_bswap(bswap_exprt &expr)
...
+    return false;
+  }
+  else if(type_id==ID_fixedbv)
+  {
+    dest.set(ID_value, integer2binary(
+      binary2integer(dest.get_string(ID_value), false)+
+      binary2integer(expr.get_string(ID_value), false),
+      unsafe_string2unsigned(dest.type().get_string(ID_width))));
```
Comment (tautschnig) [inline src/util/simplify_expr_int.cpp]: As above.

File: src/util/expr.cpp [hunk 4]
```diff
@@ -9,19 +9,20 @@ Author: Daniel Kroening, kroening@kroening.com
...
 #include "invariant.h"
-#include "expr_iterator.h"
+#include "mp_arith.h"
 #include "rational.h"
 #include "rational_tools.h"
-#include "arith_tools.h"
 #include "std_expr.h"
+#include "string2int.h"
+
+#include <cassert>
```
Comment (tautschnig) [inline src/util/expr.cpp]: I don't think this header is needed anymore.

**Review by tautschnig — COMMENTED** [review]
Minor changes, otherwise looks ok.

File: src/util/simplify_expr_int.cpp [hunk 5]
```diff
@@ -53,6 +54,114 @@ bool simplify_exprt::simplify_bswap(bswap_exprt &expr)
   return true;
 }
 
+static bool sum_expr(exprt &dest, const exprt &expr)
```
Comment (smowton) [inline src/util/simplify_expr_int.cpp]: Use the new conventionally-spaced (`a == b` not `a==b`) style?

Comment (smowton) [inline src/util/simplify_expr_int.cpp]: Add doxy comment, e.g. does this intend to add typecasts if needed, can it do pointer arithmetic, ...

File: src/util/simplify_expr_int.cpp [hunk 6]
```diff
@@ -54,47 +54,47 @@ bool simplify_exprt::simplify_bswap(bswap_exprt &expr)
   return true;
 }
 
-static bool sum_expr(exprt &dest, const exprt &expr)
+//! produce a sum of two constant expressions of the same type
+//! \return 'true' iff this was successful
```
Comment (tautschnig) [inline src/util/simplify_expr_int.cpp]: That comment is actually wrong... It might be a good case to start using enums or `optionalt` instead.

File: src/util/simplify_expr_int.cpp [hunk 7]
```diff
@@ -108,40 +108,40 @@ static bool sum_expr(exprt &dest, const exprt &expr)
   return true;
 }
 
-static bool mul_expr(exprt &dest, const exprt &expr)
+//! produce a product of two expressions of the same type
+//! \return 'true' iff this was successful
```
Comment (tautschnig) [inline src/util/simplify_expr_int.cpp]: Wrong, as above.

**Review by tautschnig — APPROVED** [review]
